### PR TITLE
Convert documentation to Python

### PR DESCRIPTION
**Pull request by jigold**
repartition.md
filtergenotypes.md
tdt.md
filtersamples_expr.md
exportgen.md
filtervariants_expr.md
persist.md
importplink.md
vep.md
write.md
annotatevariants_expr.md

### CODE REVIEW

**Comment by jigold** [comment]
@cseed back to you.

File: python/pyhail/context.py [hunk 1]
```diff
@@ -278,6 +278,61 @@ def import_plink(self, bed, bim, fam, npartitions=None, delimiter='\\\\s+', miss
         """
         Import PLINK binary file (.bed, .bim, .fam) as VariantDataset
 
+        **Examples**
+
+        Import data from a PLINK binary file
```
Comment (cseed) [inline python/pyhail/context.py]: End these lines with colons.

File: python/pyhail/context.py [hunk 2]
```diff
@@ -278,6 +278,61 @@ def import_plink(self, bed, bim, fam, npartitions=None, delimiter='\\\\s+', miss
...
+
+        >>> vds = (hc.import_plink(bed="data/test.bed",
+        >>>                        bim="data/test.bim",
+        >>>                        fam="data/test.fam"))
+
+        Phenotype is quantitative (column 6 in the .fam file)
```
Comment (cseed) [inline python/pyhail/context.py]: I'd make the description self-contained, and not reply on the previous example.  I think just one of these examples is enough.

File: python/pyhail/context.py [hunk 3]
```diff
@@ -278,6 +278,61 @@ def import_plink(self, bed, bim, fam, npartitions=None, delimiter='\\\\s+', miss
...
+        Missing value is *-9*
+
+        >>> vds = (hc.import_plink(bed="data/test.bed",
+        >>>                        bim="data/test.bim",
+        >>>                        fam="data/test.fam",
+        >>>                        missing='-9'))
+
+        **Implementation Details**
+
+        - Only binary SNP-major mode files can be read into Hail. To convert your file from Individual-major mode to SNP-major mode, use PLINK to read in your fileset and use the ``--make-bed`` option.
```
Comment (cseed) [inline python/pyhail/context.py]: Why bullets?  Just paragraphs seems fine.

File: python/pyhail/context.py [hunk 4]
```diff
@@ -278,6 +278,61 @@ def import_plink(self, bed, bim, fam, npartitions=None, delimiter='\\\\s+', miss
...
+
+        - The centiMorgan position is not currently used in Hail (Column 3 in .bim file).
+
+        - The ID (*s.id*) used by Hail is the individual ID (column 2 in .fam file).
```
Comment (cseed) [inline python/pyhail/context.py]: I think we've standardized on FAM file rather than .fam file.

File: python/pyhail/context.py [hunk 5]
```diff
@@ -278,6 +278,61 @@ def import_plink(self, bed, bim, fam, npartitions=None, delimiter='\\\\s+', miss
...
+          - 23 => "X"
+          - 24 => "Y"
+          - 25 => "X"
+          - 26 => "MT"
+
+        **Annotations**
+
+        :py:meth:`~pyhail.HailContext.import_plink` adds the following annotations:
+
+         - **va.rsid** (*String*) -- Column 2 in the *bim* file
```
Comment (cseed) [inline python/pyhail/context.py]: End with period to be consistent.

File: python/pyhail/context.py (hunk 4 above)
Comment (cseed) [inline python/pyhail/context.py]: This needs to be changed all over.

Comment (cseed) [inline python/pyhail/context.py]: Put `s.id` in double-back-ticks.

File: python/pyhail/context.py (hunk 3 above)
Comment (cseed) [inline python/pyhail/context.py]: Should "Individual" be capitalized?

File: python/pyhail/dataset.py [hunk 6]
```diff
@@ -382,7 +382,56 @@ def annotate_variants_bed(self, input, root, all=False):
     def annotate_variants_expr(self, condition):
         """Annotate variants with expression.
 
-        :param str condition: Annotation expression.
+        **Examples**
+
+        Compute GQ statistics about heterozygotes per variant
+
+        >>> (vds.annotate_variants_expr('va.gqHetStats = '
```
Comment (cseed) [inline python/pyhail/dataset.py]: I was assuming `hc` was in scope, but that's it.  Should we also assume `vds = hc.read('data/example.vds')` is in scope, too?

File: python/pyhail/dataset.py [hunk 7]
```diff
@@ -382,7 +382,56 @@ def annotate_variants_bed(self, input, root, all=False):
...
+        >>>                                'gs.filter(g => g.isHet).map(g => g.gq).stats()'))
+
+        Collect a list of sample IDs with non-ref calls in LOF variants
+
+        >>> (vds.annotate_variants_expr('va.nonRefSamples = gs.filter(g => g.isCalledNonRef).map(g => s.id).collect()'))
+
+        **Details**
```
Comment (cseed) [inline python/pyhail/dataset.py]: We should "Details", "Implementation Details", and "Notes", variously.  I vote for Notes everywhere.  Thoughts?

File: python/pyhail/dataset.py [hunk 8]
```diff
@@ -382,7 +382,56 @@ def annotate_variants_bed(self, input, root, all=False):
...
+
+        To add new variant annotations to the dataset
+        computed from existing annotations, use a named annotation
+        expression where the left hand side specifies the location
+        where the new variant annotation should be stored and the
+        right hand side specifies how to compute the new annotation
+        based on existing annotations.
+
+        For example, the annotation
+        expression ``va.isSingleton = va.info.AC == 1`` specifies that a
```
Comment (cseed) [inline python/pyhail/dataset.py]: Isn't AC an array?  Per alt alelle?

File: python/pyhail/dataset.py [hunk 9]
```diff
@@ -382,7 +382,56 @@ def annotate_variants_bed(self, input, root, all=False):
...
+
+          - ``v`` (*Variant*): :ref:`variant`
+          - ``va``: variant annotations
+          - ``global``: global annotations
+          - ``gs`` (*Aggregable[Genotype]*): aggregable of :ref:`genotype` for variant ``v``
+
+        For more information, see the documentation on writing `expressions <../overview.html#expressions>`_
+        and using the `Hail Expression Language <../reference.html#HailExpressionLanguage>`_.
+
+        Multiple variant annotations can be inserted simultaneously by specifying a list of annotation expressions to the ``condition`` argument such as ``['va.counts.nHet = gs.map(g => g.isHet).count()', 'va.counts.nHomRef = gs.map(g => g.isHomRef).count()']``.
```
Comment (cseed) [inline python/pyhail/dataset.py]: The python list is a bit long, might look better broken into two lines.

File: python/pyhail/dataset.py [hunk 10]
```diff
@@ -564,6 +613,39 @@ def downsample_variants(self, keep):
     def export_gen(self, output):
         """Export dataset as .gen file.
 
+        **Examples**
+
+        Import dosage data, filter variants based on INFO score, and export data to a .gen and .sample file
```
Comment (cseed) [inline python/pyhail/dataset.py]: end with colon:

File: python/pyhail/dataset.py [hunk 11]
```diff
@@ -564,6 +613,39 @@ def downsample_variants(self, keep):
...
+
+        >>> (vds.index_bgen("data/example.bgen").import_bgen("data/example.bgen", sample_file="data/example.sample")
```
Comment (cseed) [inline python/pyhail/dataset.py]: index_bgen is on hc, not vds.

File: python/pyhail/dataset.py [hunk 12]
```diff
@@ -564,6 +613,39 @@ def downsample_variants(self, keep):
...
+
+        **Implementation Details**
+
+        Writes out the internal VDS to a {.gen, .sample} fileset in the `Oxford spec <http://www.stats.ox.ac.uk/%7Emarchini/software/gwas/file_format.html>`_.
+
+        The first 6 columns of the resulting .gen file are the following:
+
+        - Chromosome (``v.contig``)
+        - Variant ID (``va.varid`` if exists, else Chromosome:Position:Ref:Alt)
+        - rsID (``va.rsid`` if exists, else ".")
```
Comment (cseed) [inline python/pyhail/dataset.py]: I'd say non-null or is defined, not exists.

File: python/pyhail/dataset.py [hunk 13]
```diff
@@ -799,9 +881,21 @@ def export_vcf(self, output, append_to_header=None, export_pp=False, parallel=Fa
     def write(self, output, overwrite=False):
         """Write as .vds file.
 
+        **Examples**
+
+        Import data from a vcf file and then write the data to a .vds file
```
Comment (cseed) [inline python/pyhail/dataset.py]: colon:.  VDS file.

File: python/pyhail/dataset.py [hunk 14]
```diff
@@ -564,6 +613,39 @@ def downsample_variants(self, keep):
...
+
+        Import dosage data, filter variants based on INFO score, and export data to a .gen and .sample file
+
+        >>> (vds.index_bgen("data/example.bgen").import_bgen("data/example.bgen", sample_file="data/example.sample")
+        >>>     .filter_variants_expr("gs.infoScore() >= 0.9")
+        >>>     .export_gen("data/example_filtered"))
+
+        **Implementation Details**
+
+        Writes out the internal VDS to a {.gen, .sample} fileset in the `Oxford spec <http://www.stats.ox.ac.uk/%7Emarchini/software/gwas/file_format.html>`_.
```
Comment (cseed) [inline python/pyhail/dataset.py]: I don't really like {}.

File: python/pyhail/dataset.py [hunk 15]
```diff
@@ -799,9 +881,21 @@ def export_vcf(self, output, append_to_header=None, export_pp=False, parallel=Fa
     def write(self, output, overwrite=False):
         """Write as .vds file.
```
Comment (cseed) [inline python/pyhail/dataset.py]: VDS file.

File: python/pyhail/dataset.py [hunk 16]
```diff
@@ -799,9 +881,21 @@ def export_vcf(self, output, append_to_header=None, export_pp=False, parallel=Fa
...
 
+        **Examples**
+
+        Import data from a vcf file and then write the data to a .vds file
+
+        >>> hc.import_vcf("data/sample.vcf.bgz").write("data/sample.vds")
+
+        Read data from a .vds file, filter samples, and write filtered data to a .vds file
```
Comment (cseed) [inline python/pyhail/dataset.py]: I don't think the second example adds much.

File: python/pyhail/dataset.py [hunk 17]
```diff
@@ -956,10 +1050,43 @@ def filter_alleles(self, condition, annotation=None, subset=True, keep=True, fil
         return self.hc.run_command(self, pargs)
 
     def filter_genotypes(self, condition, keep=True):
-        """Filter variants based on expression.
+        """Filter genotypes based on expression.
+
+        **Examples**
+
+        Filter genotypes by allele balance dependent on genotype call
```
Comment (cseed) [inline python/pyhail/dataset.py]: colon:.

File: python/pyhail/dataset.py [hunk 18]
```diff
@@ -956,10 +1050,43 @@ def filter_alleles(self, condition, annotation=None, subset=True, keep=True, fil
...
+
+        >>> (vds.filter_genotypes('let ab = g.ad[1] / g.ad.sum in'
+        >>>                      '((g.isHomRef && ab <= 0.1) ||'
+        >>>                      '(g.isHet && ab >= 0.25 && ab <= 0.75) ||'
+        >>>                      '(g.isHomVar && ab >= 0.9))'))
+
+        **Implementation Details**
+
+        Use the `Hail expression language <../reference.html#HailExpressionLanguage>`_ to supply a ``condition`` (boolean expression) involving the following exposed data structures:
```
Comment (cseed) [inline python/pyhail/dataset.py]: I'd use the same "genotype context so the following symbols are in scope" language.

File: python/pyhail/dataset.py [hunk 19]
```diff
@@ -956,10 +1050,43 @@ def filter_alleles(self, condition, annotation=None, subset=True, keep=True, fil
...
+        - ``global``: global annotations
+
+        All variables and values are case sensitive.
+
+        For more information, see the documentation on `data representation, annotations <../overview.html#>`_, and
+        the `expression language <../reference.html#HailExpressionLanguage>`_.
+
+        .. caution::
+
+            Missing values will always be **excluded**, regardless of whether ``keep=True`` or ``keep=False``.  Expressions in which any value is missing will evaluate to missing.
```
Comment (cseed) [inline python/pyhail/dataset.py]: This a bit unclear.  If the condition is missing, the genotype will always be excluded.

File: python/pyhail/dataset.py [hunk 20]
```diff
@@ -956,10 +1050,43 @@ def filter_alleles(self, condition, annotation=None, subset=True, keep=True, fil
...
+
+        - ``s`` (*Sample*): :ref:`sample`
+        - ``v`` (*Variant*): :ref:`variant`
+        - ``sa``: sample annotations
+        - ``va``: variant annotations
+        - ``global``: global annotations
+
+        All variables and values are case sensitive.
+
+        For more information, see the documentation on `data representation, annotations <../overview.html#>`_, and
```
Comment (cseed) [inline python/pyhail/dataset.py]: This duplicates link on the expression language above.

File: python/pyhail/dataset.py [hunk 21]
```diff
@@ -956,10 +1050,43 @@ def filter_alleles(self, condition, annotation=None, subset=True, keep=True, fil
...
+
+        Use the `Hail expression language <../reference.html#HailExpressionLanguage>`_ to supply a ``condition`` (boolean expression) involving the following exposed data structures:
+
+        - ``s`` (*Sample*): :ref:`sample`
+        - ``v`` (*Variant*): :ref:`variant`
+        - ``sa``: sample annotations
+        - ``va``: variant annotations
+        - ``global``: global annotations
+
+        All variables and values are case sensitive.
```
Comment (cseed) [inline python/pyhail/dataset.py]: This appears in a number of places.  Should we do that, or just make this remark once in the Hail expression language section?

File: python/pyhail/dataset.py [hunk 22]
```diff
@@ -986,9 +1113,56 @@ def filter_samples_all(self):
...
+        >>> (vds.annotate_samples_expr(['sa.cohort = let cohorts = ["A", "B", "C"]'
+        >>>                            'in cohorts[runif(0, cohorts.length).toInt]',
+        >>>                            'sa.isCase = pcoin(0.5)'])
+        >>>     .filter_samples_expr('(sa.cohort == "A" && sa.isCase) ||'
+        >>>                          '(sa.cohort == "B" && !sa.isCase) ||'
+        >>>                          '(sa.cohort == "C")'))
+
+        **Implementation Details**
+
+        Use the `Hail expression language <../reference.html#HailExpressionLanguage>`_ to supply a ``condition`` (boolean expression) involving the following exposed data structures:
```
Comment (cseed) [inline python/pyhail/dataset.py]: "in variant scope so blah blah".

File: python/pyhail/dataset.py [hunk 23]
```diff
@@ -1034,9 +1208,55 @@ def filter_variants_all(self):
     def filter_variants_expr(self, condition, keep=True):
         """Filter variants based on expression.
 
+        **Examples**
+
+        Keep variants on autosomal chromosomes
```
Comment (cseed) [inline python/pyhail/dataset.py]: colon

File: python/pyhail/dataset.py [hunk 24]
```diff
@@ -1034,9 +1208,55 @@ def filter_variants_all(self):
...
+        Remove variants on chromosome 1
+
+        >>> vds_filtered = (vds.filter_variants_expr('v.contig == "1"',
+        >>>                                          keep=False))
+
+        Keep variants on chromosome 21 and chromosome 22
+
+        >>> vds_filtered = (vds.filter_variants_expr('v.contig == "21" && v.contig == "22"'))
+
+        Keep variants in the gene CHD8 (assumes the variant annotation *va.gene* exists)
```
Comment (cseed) [inline python/pyhail/dataset.py]: put `va.gene` in double-back-ticks.

File: python/pyhail/dataset.py [hunk 25]
```diff
@@ -1034,9 +1208,55 @@ def filter_variants_all(self):
...
+
+        >>> vds_filtered = vds.filter_variants_expr('va.gene == "CHD8"')
+
+        Keep variants with a minor allele frequency less than 5%
+
+        >>> vds_filtered = (vds
```
Comment (cseed) [inline python/pyhail/dataset.py]: Might rewrite this one in terms of `callStats` or just use info fields.

File: python/pyhail/dataset.py [hunk 26]
```diff
@@ -1034,9 +1208,55 @@ def filter_variants_all(self):
...
+        >>>     .filter_variants_expr('let AC = gs.map(g => g.nNonRefAlleles).sum()'
+        >>>                           'and AN = gs.filter(g => g.isCalled).count()'
+        >>>                           'in AC.toDouble / AN < 0.05'))
+
+        **Implementation Details**
+
+        Use the `Hail expression language <../reference.html#HailExpressionLanguage>`_ to supply a ``condition`` (boolean expression) involving the following exposed data structures:
```
Comment (cseed) [inline python/pyhail/dataset.py]: "in sample context so ...".

File: python/pyhail/dataset.py [hunk 27]
```diff
@@ -1367,11 +1587,33 @@ def pca(self, scores, loadings=None, eigenvalues=None, k=10, arrays=False):
     def persist(self, storage_level="MEMORY_AND_DISK"):
         """Persist the current dataset.
 
+        **Examples**
+
+        Persist the dataset to both memory and disk
```
Comment (cseed) [inline python/pyhail/dataset.py]: colon.

File: python/pyhail/dataset.py [hunk 28]
```diff
@@ -1424,10 +1666,24 @@ def repartition(self, npartition, shuffle=True):
         """Increase or decrease the dataset sharding.  Can improve performance
         after large filters.
 
+        **Examples**
+
+        Increase the number of partitions by 10%
```
Comment (cseed) [inline python/pyhail/dataset.py]: I'd have one example with just a fixed number (most common use case).

File: python/pyhail/dataset.py [hunk 29]
```diff
@@ -1620,10 +1876,96 @@ def tdt(self, fam, root='va.tdt'):
         """Find transmitted and untransmitted variants; count per variant and
         nuclear family.
 
+        **Examples**
+
+        Compute TDT association results
+
+        >>> vds.tdt("data/sample.fam").export_variants("Variant = v, va.tdt.*")
```
Comment (cseed) [inline python/pyhail/dataset.py]: I like putting different commands on separate lines.  (This applies in a few other places.)

File: python/pyhail/dataset.py [hunk 30]
```diff
@@ -1620,10 +1876,96 @@ def tdt(self, fam, root='va.tdt'):
...
+
+        **Implementation Details**
+
+        The transmission disequilibrium test tracks the number of times the alternate allele is transmitted (t) or not transmitted (u) from a heterozgyous parent to an affected child under the null that the rate of such transmissions is :math:`0.5`.  For variants where transmission is guaranteed (i.e., the Y chromosome, mitochondria, and paternal chromosome X variants outside of the PAR), the test cannot be used.
```
Comment (cseed) [inline python/pyhail/dataset.py]: Math on 0.5 here seems excessive.

File: python/pyhail/dataset.py [hunk 31]
```diff
@@ -1620,10 +1876,96 @@ def tdt(self, fam, root='va.tdt'):
...
+
+        :py:meth:`~pyhail.VariantDataset.tdt` adds the following annotations:
+
+         - **tdt.nTransmitted** (*Int*) -- Number of transmitted alternate alleles.
+
+         - **va.tdt.nUntransmitted** (*Int*) -- Number of untransmitted alternate alleles.
+
+         - **va.tdt.chi2** (*Double*) -- TDT statistic.
+
+         - **va.tdt.pval** (*Double*) -- :math:`p`-value.
```
Comment (cseed) [inline python/pyhail/dataset.py]: I killed the `:math:` on p-value.

File: python/pyhail/dataset.py [hunk 32]
```diff
@@ -1620,10 +1876,96 @@ def tdt(self, fam, root='va.tdt'):
...
+
         :param str fam: Path to .fam file.
```
Comment (cseed) [inline python/pyhail/dataset.py]: FAM file

File: python/pyhail/dataset.py [hunk 33]
```diff
@@ -1644,21 +1986,227 @@ def variant_qc(self):
         pargs = ['variantqc']
         return self.hc.run_command(self, pargs)
 
-    def vep(self, config, block_size=None, root=None, force=False, csq=False):
+    def vep(self, config, block_size=1000, root="va.vep", force=False, csq=False):
```
Comment (cseed) [inline python/pyhail/dataset.py]: `'va.vep'`.  Single quotes for code and code-like strings, double-quotes "" for human readable things (error messages, documentation, etc.)

File: python/pyhail/dataset.py [hunk 34]
```diff
@@ -862,17 +923,19 @@ def export_plink(self, output, fam_expr = 'id = s.id'):
...
 
         except:
 
-        - The order among split mutli-allelic alternatives in the BED
+        - The order among split multi-allelic alternatives in the BED
           file may disagree.
         - PLINK uses the rsID for the BIM file ID.
 
-        :param str output: Output file base.  Will write BED, BIM and FAM files.
+        :param str output: Output file base.  Will write .bed, .bim, and .fam files.
```
Comment (cseed) [inline python/pyhail/dataset.py]: BED, BIM, FAM?

File: python/pyhail/dataset.py [hunk 35]
```diff
@@ -765,6 +795,40 @@ def downsample_variants(self, keep):
...
+        The first 6 columns of the resulting GEN file are the following:
+
+        - Chromosome (``v.contig``)
+        - Variant ID (``va.varid`` if defined, else Chromosome:Position:Ref:Alt)
+        - rsID (``va.rsid`` if defined, else ".")
+        - position (``v.start``)
+        - reference allele (``v.ref``)
+        - alternate allele (``v.alt``)
+
+        Probability Dosages:
```
Comment (cseed) [inline python/pyhail/dataset.py]: Why is Dosages capitalized?

File: python/pyhail/dataset.py [hunk 36]
```diff
@@ -1776,10 +1963,20 @@ def repartition(self, npartition, shuffle=True):
         """Increase or decrease the dataset sharding.  Can improve performance
         after large filters.
 
+        **Examples**
+
+        Decrease the number of partitions by 50%
```
Comment (cseed) [inline python/pyhail/dataset.py]: Can we have a repartition to fixed size?  I think that's the most common use case.

File: python/pyhail/dataset.py [hunk 37]
```diff
@@ -1972,10 +2169,98 @@ def tdt(self, fam, root='va.tdt'):
...
+
+        The transmission disequilibrium test tracks the number of times the alternate allele is transmitted (t) or not transmitted (u) from a heterozgyous parent to an affected child under the null that the rate of such transmissions is 0.5.  For variants where transmission is guaranteed (i.e., the Y chromosome, mitochondria, and paternal chromosome X variants outside of the PAR), the test cannot be used.
+
+        The TDT statistic is given by
+
+        .. math::
+
+            (t-u)^2 \over (t+u)
+
+        and follows a 1 d.o.f. chi-squared distribution under the null hypothesis.
```
Comment (cseed) [inline python/pyhail/dataset.py]: Should we spell out degree of freedom?

**Pull request merged**